rust_src/src/dired_unix.rs

```diff
@@ -316,20 +317,27 @@ fn directory_files_core(dr: &DirReq, dd: &mut DirData) -> LispObject {
 }
 
 pub fn directory_files_intro(
-    directory: LispObject,
+    directory: LispStringRef,
     full: LispObject,
     match_re: LispObject,
     nosort: LispObject,
 ) -> LispObject {
-    let dnexp = unsafe { Fexpand_file_name(directory, Qnil) };
+    let dnexp = expand_file_name(directory, None);
 
-    let handler = unsafe { Ffind_file_name_handler(dnexp, Qdirectory_files) };
+    let handler = find_file_name_handler(dnexp, Qdirectory_files);
     if handler.is_not_nil() {
-        return call!(handler, Qdirectory_files, dnexp, full, match_re, nosort);
+        return call!(
+            handler,
+            Qdirectory_files,
+            dnexp.into(),
+            full,
+            match_re,
+            nosort
+        );
     }
 
     let dr = DirReq::new(
-        dnexp.to_stdstring(),
+        LispObject::from(dnexp).to_stdstring(),
```

Comment:
This is kinda silly. We are converting a LispStringRef to a LispObject to a LispStringRef to a Rust string.

`LispObjExt` should be implemented for LispStringRef.